src/Akeneo/Channel/back/Infrastructure/EventListener/ChannelLocaleSubscriber.php

```diff
@@ -58,89 +41,59 @@ public function __construct(
     public static function getSubscribedEvents(): array
     {
         return [
-            StorageEvents::PRE_REMOVE => 'removeChannel',
-            StorageEvents::PRE_SAVE   => 'updateChannel',
+            StorageEvents::PRE_SAVE => 'storeUpdatedLocales',
+            StorageEvents::POST_SAVE => 'saveLocales',
         ];
     }
 
-    public function removeChannel(GenericEvent $event)
+    public function storeUpdatedLocales(GenericEvent $event): void
     {
         $channel = $event->getSubject();
-
         if (!$channel instanceof ChannelInterface) {
             return;
         }
 
-        $locales = $channel->getLocales();
-        $updatedLocales = [];
-
-        foreach ($locales as $locale) {
-            $locale->removeChannel($channel);
-            $updatedLocales[] = $locale;
+        foreach ($channel->getLocales() as $locale) {
+            if (!$locale->hasChannel($channel)) {
+                $locale->addChannel($channel);
+            }
+            $this->updatedLocales[$channel->getCode()][$locale->getCode()] = $locale;
         }
 
-        if (!empty($updatedLocales)) {
-            $this->saver->saveAll($updatedLocales);
+        $oldLocales = $this->repository->getDeletedLocalesForChannel($channel);
+        foreach ($oldLocales as $locale) {
+            if ($locale->hasChannel($channel)) {
+                $locale->removeChannel($channel);
+            }
+            $this->updatedLocales[$channel->getCode()][$locale->getCode()] = $locale;
+            $this->localesRemovedFromChannel[$channel->getCode()][$locale->getCode()] = $locale->getCode();
```

Comment:
What's the point to have this structure and not

```suggestion
            $this->localesRemovedFromChannel[$channel->getCode()][] = $locale->getCode();
```

Reply:
It's in order not to have several times the same locale